From d3eba503c2cee30776e4af495111a270824edd08 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Mon, 29 May 2017 16:02:00 +0300
Subject: [PATCH 0725/1200] arm64: mvebu: a8k: Turn off regulators on early
 board init

Scan the device tree for regulators that should be powered
off during boot stage and turn them off.
This action is done before COMPHY probe since some USB devices
requre that VBUS was turned OFF prior to SERDES initialization

Change-Id: I7554c5f9dace2e8511ebffaf883873eab68eabbc
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
Signed-off-by: Ken Ma <make@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/56304
Reviewed-by: Igal Liberman <igall@marvell.com>
Tested-by: Igal Liberman <igall@marvell.com>
---
 arch/arm/mach-mvebu/arm64-common.c    |  1 +
 board/Marvell/mvebu_armada-8k/board.c | 11 +++++++++++
 include/configs/mvebu_armada-8k.h     |  2 ++
 3 files changed, 14 insertions(+)

diff --git a/arch/arm/mach-mvebu/arm64-common.c b/arch/arm/mach-mvebu/arm64-common.c
index 3c84043a2c..29a7747d65 100644
--- a/arch/arm/mach-mvebu/arm64-common.c
+++ b/arch/arm/mach-mvebu/arm64-common.c
@@ -14,6 +14,7 @@
 #include <asm/arch/cpu.h>
 #include <asm/arch/soc.h>
 #include <asm/armv8/mmu.h>
+#include <power/regulator.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
diff --git a/board/Marvell/mvebu_armada-8k/board.c b/board/Marvell/mvebu_armada-8k/board.c
index 712e0ebcae..26bb19ad3f 100644
--- a/board/Marvell/mvebu_armada-8k/board.c
+++ b/board/Marvell/mvebu_armada-8k/board.c
@@ -10,6 +10,7 @@
 #include <asm/io.h>
 #include <asm/arch/cpu.h>
 #include <asm/arch/soc.h>
+#include <power/regulator.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -20,6 +21,16 @@ int board_early_init_f(void)
 	return 0;
 }
 
+int board_early_init_r(void)
+{
+#ifdef CONFIG_DM_REGULATOR
+	/* Check if any existing regulator should be turned down */
+	regulators_enable_boot_off(false);
+#endif
+
+	return 0;
+}
+
 int board_init(void)
 {
 	/* adress of boot parameters */
diff --git a/include/configs/mvebu_armada-8k.h b/include/configs/mvebu_armada-8k.h
index 56f96cc9dd..56d4c3f968 100644
--- a/include/configs/mvebu_armada-8k.h
+++ b/include/configs/mvebu_armada-8k.h
@@ -19,6 +19,8 @@
 
 #define CONFIG_NR_DRAM_BANKS	1
 
+#define CONFIG_BOARD_EARLY_INIT_R
+
 #define CONFIG_SYS_MAX_NAND_DEVICE	1
 #define CONFIG_SYS_NAND_MAX_CHIPS	1
 #define CONFIG_SYS_NAND_ONFI_DETECTION
-- 
2.22.0

